xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -170,46 +182,140 @@ public void handleNameResolutionError(Status error) {}
-              if (xdsClient == null) {
-                ManagedChannel channel = initLbChannel(helper, newBalancerName, channelCredsList);
-                xdsClient  = new XdsComms2(
-                    channel, helper, new ExponentialBackoffPolicy.Provider(),
-                    GrpcUtil.STOPWATCH_SUPPLIER, node);
-                localityStore = localityStoreFactory.newLocalityStore(helper, lbRegistry);
+              if (xdsClientRef != null) {
+                checkState(
+                    xdsClient != null, "Bug: xdsClient = xdsClientRef.getObject() not called");
+                return;
+              }
+              checkState(
+                  xdsClient == null, "Bug: xdsClient is not null while xdsClientRef is null.");
+
+              localityStore = localityStoreFactory.newLocalityStore(helper, lbRegistry);
+              final LoadReportCallback lrsCallback =
+                  new LoadReportCallback() {
+                    @Override
+                    public void onReportResponse(long reportIntervalNano) {
+                      localityStore.updateOobMetricsReportInterval(reportIntervalNano);
+                    }
+                  };
+
+              if (newXdsConfig.balancerName != null) {
+                // The is to handle the legacy usecase that requires balancerName from xds config.
+
+                String oldBalancerName = oldXdsConfig == null ? null : oldXdsConfig.balancerName;
+                if (!newXdsConfig.balancerName.equals(oldBalancerName)) {
+                  final ManagedChannel channel = initLbChannel(
+                      helper, newXdsConfig.balancerName, Collections.<ChannelCreds>emptyList());
+                  final Node node = Node.newBuilder()
+                      .setMetadata(Struct.newBuilder()
+                          .putFields(
+                              "endpoints_required",
+                              Value.newBuilder().setBoolValue(true).build()))
+                      .build();
+                  // TODO(zdapeng): Use XdsClient to do Lrs directly.
+                  lrsClient = loadReportClientFactory.createLoadReportClient(
+                      channel, helper, new ExponentialBackoffPolicy.Provider(),
+                      localityStore.getLoadStatsStore());
+                  xdsClientRef = new RefCountedXdsClientObjectPool(new XdsClientFactory() {
+                    @Override
+                    XdsClient createXdsClient() {
+                      return new XdsComms2(
+                          channel, helper, new ExponentialBackoffPolicy.Provider(),
+                          GrpcUtil.STOPWATCH_SUPPLIER, node);
+                    }
+                  });
+                } else {
+                  lrsClient = cachedLrsClient;
+                  xdsClientRef = cachedXdsClientRef;
+                }
+              } else if (cachedXdsClientRef == null) {
+                // This is to handle EDS-only with bootstrap usecase.
```

Comment:
Can you write comments more descriptively? I believe only you and me can understand what this sentence is talking about.

Reply:
Rephrased.